File: app/src/main/java/com/dani_chuks/andeladeveloper/starwars/data/db/local/dao/SpecieDao.java
package com.dani_chuks.andeladeveloper.starwars.data.db.local.dao;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Specie;

import java.util.List;

import io.reactivex.Flowable;

@Dao
public interface SpecieDao {
    @Query("SELECT * FROM specie")
    Flowable<List<Specie>> getAll();

    @Query("SELECT * FROM specie LIMIT :size")
    Flowable<List<Specie>> getItemBySize(int size);

    @Query("SELECT * FROM specie ORDER BY name ASC" )
    Flowable<List<Specie>> getAllAlphabetically();

    @Query("SELECT * FROM specie WHERE name = :name")
    Flowable <List<Specie>> getSpeciesByName(String name);

    @Query("SELECT * FROM specie WHERE url = :specieUrl")
    Flowable<Specie> getSpecieByUrl(String specieUrl);

    @Update(onConflict = OnConflictStrategy.REPLACE)
    void updateSpecies(List<Specie> specieList);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    void insertSpecieList(List<Specie> specieList);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    void insertSpecie(Specie specie);

    @Update(onConflict = OnConflictStrategy.REPLACE)
    void updateSpecie(Specie specie);
}

File: app/src/main/java/com/dani_chuks/andeladeveloper/starwars/data/db/local/dao/SpecieDao.kt
package com.dani_chuks.andeladeveloper.starwars.data.db.local.dao

import androidx.room.Dao
import androidx.room.Query
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Specie
import kotlinx.coroutines.flow.Flow

@Dao
abstract class SpecieDao: BaseDao<Specie> {
    @Query("SELECT * FROM specie")
    abstract suspend fun all(): List<Specie>

    @Query("SELECT * FROM specie ORDER BY name ASC")
    abstract suspend fun allAlphabetically(): List<Specie>

    @Query("SELECT * FROM specie LIMIT :size")
    abstract suspend fun getItemBySize(size: Int): List<Specie>

    @Query("SELECT * FROM specie WHERE name = :name")
    abstract suspend fun getSpeciesByName(name: String): List<Specie>

    @Query("SELECT * FROM specie WHERE url = :specieUrl")
    abstract suspend fun getSpecieByUrl(specieUrl: String): Specie?

    @Query("SELECT * FROM specie")
    abstract fun allAsFlow(): Flow<List<Specie>?>

    @Query("SELECT * FROM specie ORDER BY name ASC")
    abstract fun allAlphabeticallyAsFlow(): Flow<List<Specie>?>

    @Query("SELECT * FROM specie LIMIT :size")
    abstract fun getItemBySizeAsFlow(size: Int): Flow<List<Specie>?>

    @Query("SELECT * FROM specie WHERE name = :name")
    abstract fun getSpeciesByNameAsFlow(name: String): Flow<List<Specie>?>

    @Query("SELECT * FROM specie WHERE url = :specieUrl")
    abstract fun getSpecieByUrlAsFlow(specieUrl: String): Flow<Specie?>
}
